synapse/config/ratelimiting.py

```diff
@@ -21,16 +21,46 @@
 from ._base import Config
 
 
+@attr.s(slots=True, frozen=True, auto_attribs=True)
 class RatelimitSettings:
-    def __init__(
-        self,
-        config: Dict[str, float],
-        defaults: Optional[Dict[str, float]] = None,
-    ):
-        defaults = defaults or {"per_second": 0.17, "burst_count": 3.0}
+    key: str
+    per_second: float
+    burst_count: int
 
-        self.per_second = config.get("per_second", defaults["per_second"])
-        self.burst_count = int(config.get("burst_count", defaults["burst_count"]))
+
+def parse_rate_limit(
```

Comment:
Could also use a class-method, which is what we usually do.

Reply:
It was my first instinct---though I shied away from it because it's mildly irritating to type classmethods correctly in an inheritance-aware way. (`typing.Self` is the one true way, but that's in Python 3.11 and I didn't want to fiddle with typing_extensions etc).

I'll quickly make this a class method and just `-> "RatelimitSettings"`, which is good enough.